open_lm/model.py

```diff
@@ -131,24 +131,38 @@ def __init__(self, layer_id, args: Params):
             else nn.Identity()
         )
 
-    def forward(self, x: torch.Tensor, is_causal=True):
-        batchsize, seqlen, _ = x.shape
+    def forward(self, x: torch.Tensor, is_causal=True, past_key_value=None, use_cache=False):
+        batchsize, q_len, _ = x.shape
         queries, keys, vals = self.in_proj(x).chunk(3, dim=-1)
 
         queries = self.q_norm(queries)
         keys = self.k_norm(keys)
 
-        queries = queries.view(batchsize, seqlen, self.n_heads, self.head_dim)
-        keys = keys.view(batchsize, seqlen, self.n_heads, self.head_dim)
-        vals = vals.view(batchsize, seqlen, self.n_heads, self.head_dim)
-
-        queries, keys, vals = self.pos_embed(queries, keys, vals)
-
-        output = self.attn_fn(queries, keys, vals, is_causal=is_causal)
+        queries = queries.view(batchsize, q_len, self.n_heads, self.head_dim)
+        keys = keys.view(batchsize, q_len, self.n_heads, self.head_dim)
+        vals = vals.view(batchsize, q_len, self.n_heads, self.head_dim)
+
+        past_length = 0 if past_key_value is None else past_key_value[0][0].shape[1]
+        queries, keys, vals = self.pos_embed(queries, keys, vals, offset=past_length)
+
+        if past_key_value is not None and use_cache:
+            keys = torch.cat([past_key_value[0], keys], dim=1)
+            vals = torch.cat([past_key_value[1], vals], dim=1)
+        if use_cache:
+            past_key_value = (keys, vals)
+
+        # If queries and keys are the same shape, we can use causal attention
+        # Otherwise, queries are expected to be the most recent tokens, and keys are the full sequence so we use full attention
```

Comment:
Should we instead expect the caller to pass is_causal=False in this case, instead of overriding is_causal here?

Reply:
I don’t think so, the model is still causal, it’s just that when only one query is passed to generate the next token it needs to attend the whole past. It is not clear what the attention function should do when there is a different number of keys and queries with is_causal=True but with the one we have it is not what we want.

Reply:
I guess I would have it error when is_causal=True and only one query is passed, but maybe that's too purist. I'm fine leaving this as is unless anyone else reading this has a strong opinion (leave a comment!)

Reply:
I think the way they [do it](https://github.com/huggingface/transformers/blob/ac5d4cf6de24b4f7fa92996e92d1d71dd5411a6a/src/transformers/models/llama/modeling_llama.py#L886-L888) is by generating an attention mask at the out-most layer and then pass it in? This is consistent with what the [original llama](https://github.com/facebookresearch/llama/blob/main/llama/model.py#L474-L489) is doing in their repo.

Reply:
So we could also do something like this? and remove all the `is_causal` arguments but put `attention_mask` all over the place

Reply:
this was solved by pushing that to the lower-level function where it makes more sense to handle it. It is now doing two things: if q sequence size is 1 but k and v are larger then it sets causal to false assuming that q is the next token (otherwise it is matched with the first kv).
If q sequence length is larger than 1 but smaller than kv then it is assumed to match the last part of the kv for the same length (instead of the first part as simply setting is_causal to true would do).